data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,650 @@
+        unsafe { from_raw_parts(self.ptr, self.len) }
+    }
+}
+
+impl AsRef<str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn as_ref(&self) -> &str {
+        // SAFETY: created from valid utf-8
+        unsafe { from_utf8_unchecked(self.as_bytes()) }
+    }
+}
+
+impl Deref for BoxedString {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl Clone for BoxedString {
+    /// Properly clone `BoxedString` into new allocation.
+    fn clone(&self) -> Self {
+        let ptr = Box::into_raw(self.as_bytes().to_owned().into_boxed_slice()).cast::<u8>();
+        let len = self.len;
+        Self { ptr, len }
+    }
+}
+
+impl From<&str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn from(value: &str) -> Self {
+        let len = value.len();
+        let boxed = value.as_bytes().to_vec().into_boxed_slice();
+        let ptr = Box::into_raw(boxed).cast::<u8>();
+        Self { ptr, len }
+    }
+}
+
+impl From<String> for BoxedString {
+    #[inline]
+    fn from(value: String) -> Self {
+        let len = value.len();
+        let boxed = value.into_bytes().into_boxed_slice();
+        let ptr = Box::into_raw(boxed).cast::<u8>();
+        Self { ptr, len }
+    }
+}
+
+impl Drop for BoxedString {
+    #[allow(unsafe_code)]
+    fn drop(&mut self) {
+        // SAFETY: created from `Box<[u8]>`.
+        unsafe {
+            Box::<[_]>::from_raw(from_raw_parts_mut(self.ptr, self.len));
+        }
+    }
+}
+
+/// `BoxedString` is `Send` because the data they
+/// reference is unaliased. Aliasing invariant is enforced by
+/// creation of `BoxedString`.
+#[allow(unsafe_code)]
+unsafe impl Send for BoxedString {}
+
+/// `BoxedString` is `Send` because the data they
+/// reference is unaliased. Aliasing invariant is enforced by
+/// creation of `BoxedString`.
+#[allow(unsafe_code)]
+unsafe impl Sync for BoxedString {}
+
+#[derive(Clone, Copy)]
+#[repr(C)]
+struct InlinedString {
+    #[cfg(target_endian = "little")]
+    payload: [u8; MAX_INLINED_STRING_LEN],
+    /// MSB is always 1 to distinguish inlined variant.  
+    len: u8,
+    #[cfg(target_endian = "big")]
+    payload: [u8; MAX_INLINED_STRING_LEN],
+}
+
+impl InlinedString {
+    #[inline]
+    const fn len(self) -> usize {
+        (self.len - 128) as usize
+    }
+
+    #[inline]
+    const fn is_inlined(self) -> bool {
+        self.len >= 128
+    }
+
+    #[inline]
+    const fn new() -> Self {
+        Self {
+            payload: [0; MAX_INLINED_STRING_LEN],
+            // Set MSB to mark inlined variant.
+            len: 128,
+        }
+    }
+}
+
+impl AsRef<str> for InlinedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn as_ref(&self) -> &str {
+        // SAFETY: created from valid utf-8.
+        unsafe { core::str::from_utf8_unchecked(&self.payload[..self.len()]) }
+    }
+}
+
+impl<'value> TryFrom<&'value str> for InlinedString {
```

Comment:
I would usually implement `FromStr` instead

Reply:
I think `TryFrom` would be better fit here.

Comparing descriptions of this traits.

[`FromStr`](https://doc.rust-lang.org/std/str/trait.FromStr.html):
> Parse a value from a string

[`TryFrom`](https://doc.rust-lang.org/stable/std/convert/trait.TryFrom.html):
> Simple and safe type conversions that may fail in a controlled way under some circumstances. It is the reciprocal of [TryInto](https://doc.rust-lang.org/stable/std/convert/trait.TryInto.html).
> This is useful when you are doing a type conversion that may trivially succeed but may also need special handling. For example, there is no way to convert an [i64](https://doc.rust-lang.org/stable/std/primitive.i64.html) into an [i32](https://doc.rust-lang.org/stable/std/primitive.i32.html) using the [From](https://doc.rust-lang.org/stable/std/convert/trait.From.html) trait, because an [i64](https://doc.rust-lang.org/stable/std/primitive.i64.html) may contain a value that an [i32](https://doc.rust-lang.org/stable/std/primitive.i32.html) cannot represent and so the conversion would lose data. This might be handled by truncating the [i64](https://doc.rust-lang.org/stable/std/primitive.i64.html) to an [i32](https://doc.rust-lang.org/stable/std/primitive.i32.html) (essentially giving the [i64](https://doc.rust-lang.org/stable/std/primitive.i64.html)’s value modulo [i32::MAX](https://doc.rust-lang.org/stable/std/primitive.i32.html#associatedconstant.MAX)) or by simply returning [i32::MAX](https://doc.rust-lang.org/stable/std/primitive.i32.html#associatedconstant.MAX), or by some other method. The [From](https://doc.rust-lang.org/stable/std/convert/trait.From.html) trait is intended for perfect conversions, so the TryFrom trait informs the programmer when a type conversion could go bad and lets them decide how to handle it.

I think that the example with the `i64 -> i32` is indicative of our situation, we want to create `InlinedString` from `String` or `&str` but we can't due to exceeding length, so we have controlled fallback option to create `BoxedString`.

Reply:
I just dislike implicit allocations

Reply:
ok, resolve then